Diggy - Prevent oil from spawning too close to other oil patches

BlueRock2 opened the pull request:
Over time I have seen several people report that some oil patches that were too close to place adjacent pumpjacks in diggy.   Although it isn't game breaking (you can still choose one or the other to tap into) it keeps getting reported as buggy behavior.    Since vanilla oil spawn generation is too infrequent even on highest frequency for diggy, I created this update for scattered_resources.lua that will prevent oil patches from spawning less than a two tile gap from another oil patch.

BlueRock2 committed:
Prevent oil from spawning too close to other oil patches.

BlueRock2 commented:
In testing I temporarily increased oil spawns 3x and marked locations that an oil spawn was prevented using a red chest in the screenshot below.
![image](https://user-images.githubusercontent.com/45720633/74610112-24d26d00-50be-11ea-825e-e1d330dc8bb0.png)

grilledham commented on map_gen/maps/diggy/feature/scattered_resources.lua:
```diff
@@ -34,6 +34,18 @@ local function get_name_by_weight(collection, sum)
+local function overcrowded_resource(surface, position, resource_name, min_gap)
+    local d = min_gap
+    local bounding_box = {{x = position.x - d + 1, y = position.y - d + 1}, {x = position.x + d, y = position.y + d}}
+    local found_resources = surface.find_entities_filtered{area = bounding_box, name = resource_name}
```
May I suggest <https://lua-api.factorio.com/latest/LuaSurface.html#LuaSurface.count_entities_filtered> instead.
Also with `limit = 1`.

grilledham reviewed (approved):
Looks good to me.

BlueRock2 committed:
Update scattered_resources.lua

Updated to use count_entities_filtered with limit instead of find_entities_filtered

BlueRock2 commented:
Updated per grilledham's suggestions.  Re-tested again with a temporary high oil spawn rate to confirm update works as expected.   Unless anyone else needs to review or approve, this should be ready to merge.

SimonFlapse commented on map_gen/maps/diggy/feature/scattered_resources.lua:
```diff
@@ -34,6 +34,18 @@ local function get_name_by_weight(collection, sum)
+        return false
+    else
+        return true
+    end
```
Not to nitpick but you could write this in a cleaner way
I'd think either of these would work `return found_resources ~= 0` or `return not (found_resources == 0)`

However if you feel the conditional statement could be expanded with more logic, just ignore my comment.

BlueRock2 replied:
Thanks for the feedback.  Updated with your suggestion.

SimonFlapse commented on map_gen/maps/diggy/feature/scattered_resources.lua:
```diff
@@ -34,6 +34,18 @@ local function get_name_by_weight(collection, sum)
 
+--overcrowded_resource returns true if one or more entity <resource_name> is found within <min_gap> tiles from <position>
+local function overcrowded_resource(surface, position, resource_name, min_gap)
+    local d = min_gap
```
It seems you are "renaming" the `min_gap` variable to `d`. I can't see the purpose of this assignment apart from keeping the following line 6 characters shorter for each reference to the variable.

BlueRock2 replied:
Yeah, was to make the one line shorter, but it wasn't all that long so took your suggestion.

SimonFlapse reviewed (approved):
LGTM. I do have some minor suggestion that I think would improve the readability.

SimonFlapse commented on map_gen/maps/diggy/feature/scattered_resources.lua:
```diff
@@ -134,14 +146,22 @@ function ScatteredResources.register(config)
 
+        local position = {x = x, y = y}
+        if resource_name == 'crude-oil' then
+            local min_gap = 2  --must be at least this many tiles between crude-oil patches
```
Sometimes we like to write the default or recommended value for variables. Take the case that if we were to enforce a larger minimum gap for diggy, but later on choose to lower the value. In this case having a comment stating the recommended or lowest value possible would be nice to have in the future.

Just my two cents.

BlueRock2 committed:
Updated with Simon's suggestions.

BlueRock2 committed:
Update scattered_resources.lua

SimonFlapse committed:
Removed trailing whitespace (Travis CI)
